cmd/bootstrap/cmd/finalize.go

```diff
@@ -663,3 +688,39 @@ func validateEpochConfig() error {
 	}
 	return nil
 }
+
+// validateOrPopulateEpochTimingConfig validates the epoch timing config flags. In case the
+// `flagDefaultEpochTargetEndTime` value has been set, the function derives the values for
+// `flagEpochTargetEndTimeRefCounter`, `flagEpochTargetEndTimeDuration`, and `flagEpochTargetEndTimeRefTimestamp`
+// from the configuration. Otherwise, it enforces that compatible values for the respective parameters have been
+// specified (and errors otherwise). Therefore, after `validateOrPopulateEpochTimingConfig` ran,
+// the targeted end time for the epoch can be computed via `epochTargetEndTime()`.
+// You can either let the tool choose default values, or specify a value for each config.
+func validateOrPopulateEpochTimingConfig() error {
+	if flagDefaultEpochTargetEndTime {
+		// No other flags may be set
+		if !(flagEpochTargetEndTimeRefTimestamp == 0 && flagEpochTargetEndTimeDuration == 0 && flagEpochTargetEndTimeRefCounter == 0) {
+			return fmt.Errorf("invalid epoch timing config: cannot specify ANY of --epoch-target-end-time-ref-counter, --epoch-target-end-time-ref-timestamp, or --epoch-target-end-time-duration if using default timing config")
+		}
+		flagEpochTargetEndTimeRefCounter = flagEpochCounter
+		flagEpochTargetEndTimeDuration = flagNumViewsInEpoch // default to 1 view/s
+		flagEpochTargetEndTimeRefTimestamp = uint64(time.Now().Unix()) + flagNumViewsInEpoch
+
+		// compute target end time for initial (root) epoch from flags: `TargetEndTime = now() + TargetDuration`
+		rootEpochTargetEndTimeUNIX := epochTargetEndTime()
+		rootEpochTargetEndTime := time.Unix(int64(rootEpochTargetEndTimeUNIX), 0)
+		log.Info().Msgf("using default epoch timing config with root epoch target end time %s, which is in %s", rootEpochTargetEndTime, time.Until(rootEpochTargetEndTime))
+	} else {
+		// All other flags must be set
+		// NOTE: it is valid for flagEpochTargetEndTimeRefCounter to be set to 0
+		if flagEpochTargetEndTimeRefTimestamp == 0 || flagEpochTargetEndTimeDuration == 0 {
+			return fmt.Errorf("invalid epoch timing config: must specify ALL of --epoch-target-end-time-ref-counter, --epoch-target-end-time-ref-timestamp, and --epoch-target-end-time-duration")
+		}
+
+		// compute target end time for initial (root) epoch from flags: `TargetEndTime = now() + TargetDuration`
```

Comment:
I don't get how is `now()` computed. Is it the same value for all nodes or different?

Reply:
It's just describing what the following line is doing:
```go
rootEpochTargetEndTimeUNIX := rootEpochTargetEndTime()
```
We're bootstrapping, so `now()` is invoked once as `time.Now()` on the machine where bootstrapping is occurring.

Reply:
Ah, this comment was outdated I think, from an earlier way I was thinking of computing this. Updated here: https://github.com/onflow/flow-go/pull/5027/commits/996334cbc7ca78919765a78a9b1ea74a18256d33